Add known_types/0 function that lists all known types

goncalotomas opened the pull request:
Adds a `known_types/0` function that returns a list of all known types. Fixes #75.

```
iex(1)> MIME.known_types
["image/bmp", "application/json-patch+json",
 "application/vnd.oasis.opendocument.spreadsheet", "application/zip",
 "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
 "application/xhtml+xml", "application/x-bzip2", "application/manifest+json",
 "application/vnd.ms-powerpoint", "application/json", "video/mpeg",
 "application/x-7z-compressed", "video/mp4",
 "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
 "text/css", "application/vnd.etsi.asic-e+zip", "text/xml", "text/plain",
 "audio/aac", "image/svg+xml", "video/x-msvideo", "image/heic", "image/avif",
 "text/html",
 "application/vnd.openxmlformats-officedocument.presentationml.presentation",
 "application/x-abiword", "application/postscript", "video/ogg", "audio/opus",
 "application/x-freearc", "audio/webm", "font/woff", "text/markdown",
 "application/x-httpd-php", "audio/wav", "application/vnd.amazon.ebook",
 "application/x-bzip", "application/rss+xml", "image/png", "application/ogg",
 "application/wasm", "application/x-cdf", "video/x-ms-wmv", "audio/midi",
 "video/webm", "video/quicktime", "image/jpeg", "image/vnd.adobe.photoshop",
 "text/calendar", "video/3gpp2", ...]
```

whatyouhide commented on lib/mime.ex:
```diff
@@ -164,6 +164,13 @@ defmodule MIME do
+  @doc """
+  Returns a list of all known types, including custom types compiled into the MIME module.
+  """
+  def known_types do
```
```suggestion
  @doc since: "2.1.0"
  @spec known_types() :: [String.t(), ...]
  def known_types do
```

goncalotomas replied:
Applied a custom version of this suggestion since the type spec changed a bit :)

whatyouhide reviewed (commented):
Nice! Can we add a test for this? It's enough to test that the returned term is a list of strings, has some known types in it, and has some user-defined known types in it.

josevalim commented on lib/mime.ex:
```diff
@@ -164,6 +164,13 @@ defmodule MIME do
+  Returns a list of all known types, including custom types compiled into the MIME module.
+  """
+  def known_types do
+    Map.keys(unquote(Macro.escape(all_types)))
```
Wouldn't it be better to compile them with their extensions, so all information becomes available?

goncalotomas replied:
I made it return just the types because that's what I would expect a `known_types` function to do but it's also better to have more information. I changed it to return the mapping of types to extensions. :)

goncalotomas commented:
Added a test and changed the return type to include extensions:
```elixir
iex(1)> MIME.known_types
%{
  "image/bmp" => ["bmp"],
  "application/json-patch+json" => ["json-patch"],
  "application/vnd.oasis.opendocument.spreadsheet" => ["ods"],
  "application/zip" => ["zip"],
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document" => ["docx"],
  "application/xhtml+xml" => ["xhtml"],
  "application/x-bzip2" => ["bz2"],
  "application/manifest+json" => ["webmanifest"],
  "application/vnd.ms-powerpoint" => ["ppt"],
  "application/json" => ["json"],
  "video/mpeg" => ["mpeg", "mpg"],
  "application/x-7z-compressed" => ["7z"],
  "video/mp4" => ["mp4"],
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" => ["xlsx"],
  "text/css" => ["css"],
  "application/vnd.etsi.asic-e+zip" => ["asice", "sce"],
  "text/xml" => ["xml"],
  "text/plain" => ["txt", "text"],
  "audio/aac" => ["aac"],
  ...
}
```

whatyouhide commented on lib/mime.ex:
```diff
@@ -164,6 +164,16 @@ defmodule MIME do
 
+  @doc """
+  Returns a mapping of all known types to their extensions,
+  including custom types compiled into the MIME module.
```
Might help visual learners 😉 

```suggestion
  including custom types compiled into the MIME module.
  
  ## Examples
  
      known_types()
      #=> %{"application/json" => ["json"], ...}

```

whatyouhide commented on lib/mime.ex:
```diff
@@ -164,6 +164,16 @@ defmodule MIME do
+  including custom types compiled into the MIME module.
+  """
+  @doc since: "2.1.0"
+  @spec known_types() :: %{String.t() => [String.t()]}
```
```suggestion
  @spec known_types() :: %{required(String.t()) => [String.t()]}
```

whatyouhide reviewed (approved):
Awesome!

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
